### ISSUE TITLE
Images are not loading in the HTML file.

### ISSUE THREAD

**Issue by ankitgupta0304** [anshooarora/extentreports#243]
using below code to add image in the HTML file.
    	String imagePath = ScreenShot.shootScreenShot(SeleniumMethds.driver, ImageAura_WE, "verifyDate");
    	ExtentReportFile.logger.log(LogStatus.INFO, "ShootImage has been caputred at :"+ imagePath);
    	String image = ExtentReportFile.logger.addScreenCapture(imagePath);
    	System.out.println("shoot image path for extent report "+image);

But images are not displaying in the file. All the images are broken.
Please suggest.

**Comment by anshooarora** [comment]
See [here](http://extentreports.relevantcodes.com/java/version2/docs.html#insert-screenshots) for the fix.

**Comment by testphreak** [comment]
Anshoo, I tried adding screenshots to the report (v2.40), but it doesn't seem to work for me. All I get is a little blank image in the report. Am I missing something obvious? I did look at the documentation to re-confirm.

```java
ExtentReporter.getTest().log(LogStatus.INFO, "Login Page: " + ExtentReporter.getTest().addScreenCapture(TestLogger.captureScreenShot(webDriver)));
```

Screen shot method:

```java
public static String captureScreenShot(final RemoteWebDriver webDriver) {
        File screenshotFile = webDriver.getScreenshotAs(OutputType.FILE);
        File targetFile = new File(screenShotsDirectory, getUniqueString() + ".png");
        try {
            FileUtils.copyFile(screenshotFile, targetFile);
        } catch (IOException e) {
            ExtentReporter.getTest().log(LogStatus.INFO, "Unable to log screenshot");
        }
        return targetFile.getAbsolutePath();
    }
```
Variable values look like this during the run:
screenshotFile - /var/folders/r4/dgdmhsk91zx31jmk4f53wfkh0000gp/T/screenshot4415488863879477779.png
screenShotsDirectory - /Users/testphreak/git/turbo/test/webdriver/test-output/html
targetFile.getAbsolutePath() - /Users/testphreak/git/turbo/test/webdriver/test-output/html/1451953514355-1.png

**Comment by anshooarora** [comment]
I tested and am unable to reproduce.  This works for me using both relative and absolute paths.  Can you remove the first '/' and try again?  You can also add the relative path using developer-tools/Firebug to see what path works.  Extent doesn't do anything special with the path, it just adds it as-is so if its relative to the report, it will link to the image.

Consider the below hierarchy:

```
Reports
0.png
    - Extent.html
    - 1.png
    - 2.png
    - Snapshots
          - 3.png
          - 4.png
```

From the perspective of Extent.html:

```
../0.png
1.png or ./1.png
2.png or ./2.png
Snapshots/3.png or ./Snapshots/3.png
Snapshots/4.png or ./Snapshots/3.png
```

**Comment by testphreak** [comment]
Thanks Anshoo. Figured out the problem. I was returning the absolute path of the file [targetFile.getAbsolutePath()], when the addScreenCapture expects just the file name as String. So I modified captureScreenShot method to return just the file name and pass it to the addScreenShot method.

```java
//I am keeping the report html file and images in the same directory (screenShotsDirectory)
public static String captureScreenShot(final RemoteWebDriver webDriver) {
        String imageName = getUniqueString() + ".png";
        File screenshotFile = webDriver.getScreenshotAs(OutputType.FILE);
        File targetFile = new File(screenShotsDirectory, imageName);
        try {
            FileUtils.copyFile(screenshotFile, targetFile);
        } catch (IOException e) {
            ExtentReporter.getTest().log(LogStatus.INFO, "Unable to log screenshot");
        }
        return targetFile.getName();
    }
```
Maybe obvious to most, but something to note here is that absolute path to image works, when you open the html file from the local system (using File --> Open or from Finder/Explorer). Absolute path to image didn't work if loading with http, for example - http://localhost:63349/turbo-git-webdriver/turbo-webdriver/test-output/html/index.html. This is because modern browsers aren't allowed to access local system resources as a security measure.

*Couple of questions/follow up -*
1. Perhaps it makes sense to modify the addScreenCapture method param to String imageName perhaps to avoid confusion and update the @param description and documentation online indicating the name needs to be supplied and not the path?
2. Is there a view called ImageCollectionView? A test I ran generated 3 images, all of which I can see inline in the test, but according to the documentation, I got the impression that there's also a separate view with all the images in one place.

**Comment by anshooarora** [comment]
1. Feel free to change the docs of ExtentTest, I will merge.  In the previous versions, this method-name made sense because the image could be added to the test as well as it was automatically added to a media view, but in the current versions, this view no longer exists.  Now, it only returns a HTML string with the image path that can be supplied to any log, but I didn't change it as user tests were going to break with this change.,

2. The docs should be corrected here as the earlier versions had a separate view to display all media used by the user.  Versions 2.10+ do not have this and the media is only viewed in the test.  This view can be added back very easily but I felt there would be too many views and too much information to process in the report.

**Comment by testphreak** [comment]
1. Updated the Javadoc. My feedback was regarding the parameter name of the addScreenCapture method and not the method name itself. So the method name is perfectly fine. Also, in my tests, what I found was that the method correctly inserts the images inline in steps if the image is in the same directory as the report file. If the images were in another directory images would appear broken in the report. I must state, that my tests were on a Mac and I was able to get the screen captures to show up running locally and using http (for example - http://localhost:63349/turbo-git-webdriver/turbo-webdriver/test-output/html/index.html#!) through this method i.e. image in the same directory as the report.

2. I agree, ImageCollectionView is redundant. I was just curious.

3. On a side note, noticed that TestNG was on 6.9.6 in the pom file. Could we update it to 6.9.9 (one below the latest version), that way any issues can be reported back to the TestNG project while the version is relatively new?

**Comment by anshooarora** [comment]
If the images are in another folder, passing the correct path would still link to the image.  I saw your docs, in that you have named the param as imageName but that is not entirely correct as the path is what gets linked:

```
// absolute:
C:\Users\Anshoo\images\1.png

// relative, drill-down
./child-folder/1.png

// relative, climb up
../parent-folder/1.png
```

Feel free to send in another pull-request for the version..

**Comment by testphreak** [comment]
You're correct that you can have the screenshots in parent or child directories in relation to where the report is. Thanks for explaining this a little more.

But in Java (at least to my knowledge) there's no way to construct the path the way you are showing it i.e. with ./child-folder/1.png and that was the gap for me reading the Javadoc. With imagePath description, I got the impression that one of either getPath() or getAbsolutePath() would work when passing it to the addScreenCapture() method. But in reality I need to manually construct the path to the screenshot *in context of* where the report file resides.

Made the following change in my project to work with screenshots and the report file being in different directories. Hopefully this will be useful to anybody else who may have the problem integrating screenshots with the report.

```java
//TestLogger.java (class container wrapper log methods)
/**
     * Capture screen shot and log using Extent Reports
     * @param webDriver driver
     * @param text text associated with screenshot
     */
    public static void captureScreenShot(String text, RemoteWebDriver webDriver) {
        String imageName = getUniqueString() + ".png";
        File screenshotFile = webDriver.getScreenshotAs(OutputType.FILE);
        File targetFile = new File(screenShotsDirectory, imageName);
        try {
            FileUtils.copyFile(screenshotFile, targetFile);
        } catch (IOException e) {
            ExtentReporter.getTest().log(LogStatus.INFO, "Unable to log screenshot");
        }
        String screenshotsPath = "./screenshots/" + targetFile.getName();
        ExtentReporter.getTest().log(LogStatus.INFO, text +     ExtentReporter.getTest().addScreenCapture(screenshotsPath));
 }
```
Then use it in my tests as such
```java
TestLogger.captureScreenShot("Screenshot of page", webDriver);
```

My Javadoc update is incorrect. Does it make sense to add a note indicating that you would need to construct the path yourself? If that's pretty obvious, then please decline the pull request and delete my branch.

**Comment by anshooarora** [comment]
Actually the methods used at runtime do not build a path that would ideally be best suited to our needs.  

Only a few minor corrections are required in your pull request, you can send in another one and I will merge.  I wouldn't want your contributions slipping away.

**Comment by testphreak** [comment]
Yeah, I agree. Hopefully the change I committed this time isn't too wordy.

**Comment by anshooarora** [comment]
Looks great.

I am closing this as the new docs provide details on the correct usage of this method.  Feel free to open a new ticket and reference this in case of issues.

**Comment by rachanarathore** [comment]
Hi anshooarora 

Is there any way where we can add expected and actual result in extent reports

**Comment by Shroet** [comment]
How can I store .png files in the report v 2.0? These files are displaying fine in a report but only on the local machine. When I'm sending this report to another person nobody sees this png files. I give them next path : C:/projectName/src/screenshots/TC-4-Login.png . Anybody know how can I solve this issue?

**Comment by Naimadnap** [comment]
@anshooarora Is something changing in displaying images in the report?
I try to add screenshot from file this way
`MediaEntityModelProvider mediaModel = MediaEntityBuilder.createScreenCaptureFromPath("InvalidImageUrls.png").build();
        report.info("details", mediaModel);`
but I get 404 in browser.
The InvalidImageUrls.png is in the same folder as ExtentReport.html according to your post 
> anshooarora commented on 5 Jan 2016
in this issue

**Comment by nguyenngocnhan** [comment]
Hi!
I have issue that the image can display  on the report but it have error:

java.io.UncheckedIOException: java.io.IOException: The system cannot find the file specified ./output/MasterImages/home-xp_Master.png
	at com.aventstack.extentreports.Report.lambda$addScreenCapture$7(Report.java:121)
the report is on \reperting\myreport.hmtl

How can I fixed it?

Thank you.

**Comment by gnpavan430** [comment]
I am using extent reports version 2.41.2. Images are not displayed in the report. They are displayed as broken images. What should I do?

My getScreenshot method is
 public static String getScreenshot(WebDriver driver, String screenshotName) throws Exception {
        String dateName = new SimpleDateFormat("yyyyMMddhhmmss").format(new Date());
        TakesScreenshot ts = (TakesScreenshot) driver;
        File source = ts.getScreenshotAs(OutputType.FILE);
        //after execution, you could see a folder "FailedTestsScreenshots" under src folder
        String destination = System.getProperty("user.dir") + "/Screenshots/"+screenshotName+dateName+".png";
        File finalDestination = new File(destination);
        FileUtils.copyFile(source, finalDestination);
        System.out.println("Destination is"+destination);
        return destination;
    }

**Comment by c5242** [comment]
My issue is not resolved , I am facing broken issue 
How I can fix it ?
public String takeScreenShot(String testName)
	{
		Date d = new Date();
		String screenshotFile = d.toString().replace(":", "_").replace(" ", "_");
		// store screenshot in that file
		String finalName = screenshotFile + "_" + testName + ".png";
		File scrFile = ((TakesScreenshot) driver).getScreenshotAs(OutputType.FILE);
		try
		{
			FileUtils.copyFile(scrFile, new File(System.getProperty("user.dir") + "//screenshots//" + finalName));
		}
		catch (IOException e)
		{
			// TODO Auto-generated catch block e.printStackTrace();
		}
		// put screenshot file in reports
		test.log(LogStatus.INFO, "Screenshot-> " + test.addScreenCapture(finalName));
		return System.getProperty("user.dir") + "//screenshots//" + finalName;
	}

![image](https://user-images.githubusercontent.com/26655284/43777145-9458ac5e-9a6f-11e8-9cd8-93709e8207c7.png)

**Comment by c5242** [comment]
I hv to use absolute path ..what could be the solution ?

**Comment by nhidinh** [comment]
In my case, as I just need to see the screenshot in report so I attached it as base64screenshot and it work. 
Code is as below: 
`public synchronized void onTestFailure(ITestResult result) {
        System.out.println("==="+methodDes + "=== failed!");
        try {
            WebDriver driver = (WebDriver) result.getTestContext().getAttribute("driver");
            String base64Screenshot = ((TakesScreenshot)driver).getScreenshotAs(OutputType.BASE64);
            MediaEntityModelProvider mediaModel = MediaEntityBuilder.createScreenCaptureFromBase64String(base64Screenshot).build();
            test.get().fail("captured image: ", mediaModel);
        } catch (IOException e) {
            e.printStackTrace();
        }
        test.get().fail(result.getThrowable().getMessage());
    }
`
